### PR TITLE
Fabric proxy model changes

### PR DESCRIPTION
**Pull request by Ram-srini**
1. Fabric sdk implementation classes
2. Fabric connector service
3. Generic client for fabric proxy model

Signed-off-by: Ramakrishna Srinivasamurthy <ramakrishna.srinivasamurthy@intel.com>

### CODE REVIEW

File: sdk/avalon_sdk/fabric/fabric_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+            with open(self.__network_conf_file, 'r') as profile:
+                self.__network_config = json.load(profile)
+            self.__channel_name = config["fabric"]["channel_name"]
+            self.__orgname = base.get_net_info(self.__network_config,
+                                               'client', 'organization')
+            logging.info("Org name choose: {}".format(self.__orgname))
+            self.__peername = random.choice(base.get_net_info(
+                self.__network_config, 'organizations', self.__orgname,
+                'peers'))
+            # Get a tx commiter
```
Comment (danintel) [inline sdk/avalon_sdk/fabric/fabric_wrapper.py]: Change "commiter" to "committer". Fabric misspelled committer here in the object, but we do not have to in the comment.

File: blockchain_connector/Dockerfile-fabric [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    pylint3 \
+    cppcheck \
+ && apt-get -y -q upgrade \
+ && apt-get clean \
+ && rm -rf /var/lib/apt/lists/*
+
+# Make Python3 default
+RUN ln -sf /usr/bin/python3 /usr/bin/python
+
+# Build Avalon blockchain connecetor intermediate docker image
```
Comment (danintel) [inline blockchain_connector/Dockerfile-fabric]: Change connecetor to connector.

File: sdk/avalon_sdk/fabric/fabric_work_order.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        else:
+            raise Exception("config is none")
+
+    def work_order_submit(self, work_order_id, worker_id, requester_id,
+                          work_order_request, id=None):
+        """
+        Submit work order request to fabric block chain.
+        Params
+            work_order_id is unique id of the work order request
+            worker_id is identifer for the worker
```
Comment (danintel) [inline sdk/avalon_sdk/fabric/fabric_work_order.py]: Change identifer to identifier

File: tests/test_encrypted_data_encryption.py [hunk 4]
```diff
@@ -39,178 +39,179 @@
...
+
+    # Sign work order
+    private_key = utility.generate_signing_keys()
+    session_key = utility.generate_key()
+    session_iv = utility.generate_iv()
+    encrypted_session_key = utility.generate_encrypted_key(
+        session_key, worker_obj.worker_encryption_key)
+    # Generate one time key used for encrypting inData
+    data_key = utility.generate_key()
+    # Initialation vector of size 12 bytes with all zeros.
```
Comment (danintel) [inline tests/test_encrypted_data_encryption.py]: Change Initialation to Initialization.

File: blockchain_connector/avalon_blockchain_connector/connector_service.py [hunk 5]
```diff
@@ -96,8 +99,12 @@ def main(args=None):
             logging.info("About to start Ethereum connector service")
             eth_connector_svc = EthereumConnector(config)
             eth_connector_svc.start()
-        else:
+        elif proxy_blockchain == FABRIC:
             logging.info("About to start Fabric connector service")
+            fabric_connector_svc = FabricConnector(uri)
+            fabric_connector_svc.start()
+        else:
```
Comment (rranjan3) [inline blockchain_connector/avalon_blockchain_connector/connector_service.py]: Can be removed. The argparser takes care of this.

File: sdk/avalon_sdk/fabric/fabric_work_order.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+            return event_handler
+        else:
+            logging.error(
+                "Fabric wrapper instance is not initialized")
+            return None
+
+    def get_work_order_completed_event_handler(self, handler_func):
+        """
+        Function to start event handler loop for
+        workOrderSubmitted event
```
Comment (rranjan3) [inline sdk/avalon_sdk/fabric/fabric_work_order.py]: You probably intended workOrderCompleted

File: docker-compose-fabric.yaml [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        - MAKECLEAN=${MAKECLEAN:-1}
+        - TCF_DEBUG_BUILD=${TCF_DEBUG_BUILD:-}
+        - DISPLAY=${DISPLAY:-}
+        - XAUTHORITY=~/.Xauthority
+    volumes:
+      # Below volume mappings are required for DISPLAY settings by heart disease GUI client
+      - /tmp/.X11-unix:/tmp/.X11-unix
+      - ~/.Xauthority:/root/.Xauthority
+      - ~/mywork/vars/keyfiles:/keyfiles
+      - ~/mywork/vars/mychannel_network.json:/project/TrustedComputeFramework/sdk/avalon_sdk/fabric/network.json
```
Comment (manojgop) [inline docker-compose-fabric.yaml]: Is the above volume mapping for keyfiles and network.json required ? Could you please add a comment

File: docker-compose-fabric.yaml [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        avalon_blockchain_connector -b fabric -u http://avalon-listener:1947
+        tail -f /dev/null
+        "
+    volumes:
+      - ./:/project/avalon
+      - ~/mywork/vars/keyfiles:/keyfiles
+      - ~/mywork/vars/mychannel_network.json:/project/TrustedComputeFramework/sdk/avalon_sdk/fabric/network.json
+    depends_on:
+      - avalon-listener
+      - tcf
```
Comment (manojgop) [inline docker-compose-fabric.yaml]: Just FYI - After enclave manager PR is merged, tcf container just consists the avalon shell. So we need to change dependency of avalon-blockchain-connector from tcf to enclave manager.

File: scripts/start_fabric.sh [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+-e "NO_PROXY=$no_proxy,localhost,orderer3.example.com,orderer2.example.com,orderer1.example.com,peer2.org1.example.com,peer1.org1.example.com,peer2.org0.example.com,peer1.org0.example.com" \
+hyperledger/fabric-tools:2.0
+echo "Installing and instantiating registry chain code.."
+minifab install,approve,instantiate -n registry -v 1.0
+echo "Installing and instantiating worker chain code.."
+minifab install,approve,instantiate -n worker -v 1.0
+echo "Installing and instantiating order chain code.."
+minifab install,approve,instantiate -n order -v 1.0
+echo "Installing and instantiating receipt chain code.."
+minifab install,approve,instantiate -n receipt -v 1.0
```
Comment (manojgop) [inline scripts/start_fabric.sh]: Please add extra line at end of file

File: sdk/avalon_sdk/direct/avalon_direct_client.py [hunk 10]
```diff
@@ -61,7 +61,7 @@ def __init__(self, config_file=None, config=None):
                     raise Exception('Could not open config file: %s' % e)
 
         self.__blockchain_type = self.__config['blockchain']['type']
-        if self.__blockchain_type == "Ethereum":
+        if self.__blockchain_type == "ethereum":
```
Comment (manojgop) [inline sdk/avalon_sdk/direct/avalon_direct_client.py]: May be do this check -> self.__blockchain_type.lower() == 'ethereum'

File: sdk/avalon_sdk/fabric/fabric_work_order_receipt.py [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+    def work_order_receipt_retrieve(self, work_order_id):
+        """
+        Retrieving a Work Order Receipt
+        Inputs:
+        work_order_id is the id of the Work Order to be retrieved.
+        id is used for json rpc request
+        Outputs:
+        worker_service_id, requester_id, work_order_id, receipt_create_status,
+        and work_order_request_hash are defined in work_order_receipt_create().
+        None on error
```
Comment (manojgop) [inline sdk/avalon_sdk/fabric/fabric_work_order_receipt.py]: The function is returning -1 on error

File: sdk/avalon_sdk/fabric/fabric_work_order_receipt.py [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+        Params
+            work_order_id is a Work Order id that was
+            sent in the corresponding work_order_submit request.
+            updater_id is an id of the updating entity. It is optional
+            update_index is an index of the update to retrieve.
+            Value "0xFFFFFFFF" is reserved to retrieve the last
+            received update.
+        Returns
+            updater_id, update_type, update_data, update_signature, and
+            signature_rules are defined work_order_receipt_update().
```
Comment (manojgop) [inline sdk/avalon_sdk/fabric/fabric_work_order_receipt.py]: The function is returning -1 or error and work_order_result otherwise. looks like inconsistent return value.

File: sdk/avalon_sdk/fabric/fabric_work_order_receipt.py [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+            params = []
+            params.append(work_order_id)
+            params.append(updater_id)
+            params.append(update_index)
+            receipt_result = self.__fabric_wrapper.invoke_chaincode(
+                self.CHAIN_CODE,
+                'workOrderReceiptUpdateRetrieve',
+                params)
+            if receipt_result is not None:
+                return work_order_result
```
Comment (manojgop) [inline sdk/avalon_sdk/fabric/fabric_work_order_receipt.py]: should it return receipt_result or work_order_result ?

File: sdk/avalon_sdk/fabric/fabric_work_order_receipt.py [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+            params.append(worker_service_id)
+            params.append(worker_id)
+            params.append(requester_id)
+            params.append(receipt_status)
+            receipt_result = self.__fabric_wrapper.invoke_chaincode(
+                self.CHAIN_CODE,
+                'workOrderReceiptLookUp',
+                params)
+            if receipt_result is not None:
+                return work_order_result
```
Comment (manojgop) [inline sdk/avalon_sdk/fabric/fabric_work_order_receipt.py]: should it return receipt_result or work_order_result ?

**Comment by Ram-srini** [comment]
> This is an impressive set of changes to support HL Fabric.
> I would like to see instructions in a README.md on how to run this (for example, using the echo or heart disease or any worker). What you sent to me separately in an email would be good for a README.md
> 
> I have minor corrections to comments here.

I have added readme file in doc.

File: sdk/avalon_sdk/fabric/fabric_wrapper.py (hunk 1 above)
Comment (Ram-srini) [reply sdk/avalon_sdk/fabric/fabric_wrapper.py]: Done

File: blockchain_connector/Dockerfile-fabric (hunk 2 above)
Comment (Ram-srini) [reply blockchain_connector/Dockerfile-fabric]: Done

File: sdk/avalon_sdk/fabric/fabric_work_order.py (hunk 3 above)
Comment (Ram-srini) [reply sdk/avalon_sdk/fabric/fabric_work_order.py]: Done

File: tests/test_encrypted_data_encryption.py (hunk 4 above)
Comment (Ram-srini) [reply tests/test_encrypted_data_encryption.py]: Done

File: blockchain_connector/avalon_blockchain_connector/connector_service.py (hunk 5 above)
Comment (Ram-srini) [reply blockchain_connector/avalon_blockchain_connector/connector_service.py]: Done.

File: sdk/avalon_sdk/fabric/fabric_work_order.py (hunk 6 above)
Comment (Ram-srini) [reply sdk/avalon_sdk/fabric/fabric_work_order.py]: Changed.

File: blockchain_connector/Dockerfile-fabric [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+RUN apt-get update \
+ && apt-get install -y -q \
+    build-essential \
+    ca-certificates \
+    cmake \
+    git \
+    make \
+    pkg-config \
+    cmake \
+    protobuf-compiler \
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply blockchain_connector/Dockerfile-fabric]: Make is required, others can be removed.

File: docker-compose-fabric.yaml [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+      bash -c "
+        avalon_blockchain_connector -b fabric -u http://avalon-listener:1947
+        tail -f /dev/null
+        "
+    volumes:
+      - ./:/project/avalon
+      # Fabric crypto materials has to be shared to connector to
+      # interact with fabric blockchain
+      - ~/mywork/vars/keyfiles:/keyfiles
+      - ~/mywork/vars/mychannel_network.json:/project/TrustedComputeFramework/sdk/avalon_sdk/fabric/network.json
```
Comment (pankajgoyal2) [inline docker-compose-fabric.yaml]: change /project/TrustedComputeFramework/sdk/ to /project/avalon/sdk/

Comment (Ram-srini) [reply docker-compose-fabric.yaml]: Fxied.

File: blockchain_connector/avalon_blockchain_connector/fabric/fabric_connector.py [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+            logging.info("get work order submitted event handler failed")
+            return None
+
+    def workorder_event_handler_func(self, event, block_num, txn_id, status):
+        logging.info("Event payload: {}\n Block number: {}\n"
+                     "Transaction id: {}\n Status {}".format(
+                         event, block_num, txn_id, status
+                     ))
+        jrpc_req_id = 301
+        # work_order_req = json.loads(event)
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply blockchain_connector/avalon_blockchain_connector/fabric/fabric_connector.py]: Done

File: blockchain_connector/avalon_blockchain_connector/fabric/fabric_connector.py [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+            work_order_req['workerId'],
+            work_order_req['requesterId'],
+            work_order_req["workOrderRequest"],
+            id=jrpc_req_id
+        )
+        logging.info("Work order submit response {}".format(
+            response
+        ))
+        if response and 'error' in response and \
+                response['error']['code'] == 5:
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply blockchain_connector/avalon_blockchain_connector/fabric/fabric_connector.py]: Changed

File: sdk/avalon_sdk/direct/avalon_direct_client.py (hunk 10 above)
Comment (Ram-srini) [reply sdk/avalon_sdk/direct/avalon_direct_client.py]: Done

File: sdk/avalon_sdk/fabric/fabric_work_order_receipt.py (hunk 12 above)
Comment (Ram-srini) [reply sdk/avalon_sdk/fabric/fabric_work_order_receipt.py]: Changed

File: sdk/avalon_sdk/fabric/fabric_work_order_receipt.py (hunk 13 above)
Comment (Ram-srini) [reply sdk/avalon_sdk/fabric/fabric_work_order_receipt.py]: Changed

File: sdk/avalon_sdk/fabric/fabric_work_order_receipt.py (hunk 14 above)
Comment (Ram-srini) [reply sdk/avalon_sdk/fabric/fabric_work_order_receipt.py]: Changed.

File: sdk/avalon_sdk/fabric/fabric_work_order_receipt.py (hunk 11 above)
Comment (Ram-srini) [reply sdk/avalon_sdk/fabric/fabric_work_order_receipt.py]: Done

File: docker-compose-fabric.yaml [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+        - MAKECLEAN=${MAKECLEAN:-1}
+        - TCF_DEBUG_BUILD=${TCF_DEBUG_BUILD:-}
+        - DISPLAY=${DISPLAY:-}
+        - XAUTHORITY=~/.Xauthority
+    volumes:
+      # Below volume mappings are required for DISPLAY settings by heart disease GUI client
+      - /tmp/.X11-unix:/tmp/.X11-unix
+      # To store the bookmark file to host machine,
+      # it is required for fabric event listening
+      - ./examples/apps/generic_client/:/project/TrustedComputeFramework/examples/apps/generic_client/
```
Comment (pankajgoyal2) [inline docker-compose-fabric.yaml]: Please replace TrustedComputeFramework with avalon at all places.

File: blockchain_connector/Dockerfile-fabric [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+   && pip3 install dist/*.whl
+
+# Build final image and install modules
+FROM base_image as final_image
+
+ENV TCF_HOME=/project/avalon
+
+WORKDIR /project/avalon/blockchain_connector
+
+COPY --from=build_image /usr/local/ /usr/local
```
[the remark this answers is not in the document]
Comment (manojgop) [reply blockchain_connector/Dockerfile-fabric]: I have raised an issue https://github.com/hyperledger/avalon/issues/290. Please raise a separate PR to fix it

File: docker/Dockerfile.tcf-dev [hunk 21]
```diff
@@ -50,7 +50,7 @@ RUN apt-get update \
     protobuf-compiler \
     libprotobuf-dev \
     python3-dev \
-    virtualenv \
+    libssl-dev \
```
Comment (manojgop) [inline docker/Dockerfile.tcf-dev]: Why we need libssl-dev ?  Is it used by Fabric ?

File: docker-compose-fabric.yaml (hunk 19 above)
Comment (Ram-srini) [reply docker-compose-fabric.yaml]: PR#266 address that.

File: examples/apps/generic_client/fabric_generic_client.py [hunk 22]
```diff
@@ -0,0 +1,633 @@
+#! /usr/bin/env python3
+
+# Copyright 2019 Intel Corporation
```
Comment (manojgop) [inline examples/apps/generic_client/fabric_generic_client.py]: 2020

File: sdk/avalon_sdk/fabric/fabric_wrapper.py [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+        """
+        This is wrapper method to invoke the chain code.
+        Params:
+            chaincode_name is name of the chain code
+            method_name is chain code method name
+            params is list of arguments to method
+        Returns:
+            If the call to chain code query then it
+            returns the payload of chain code response
+            on success or None on error.
```
Comment (manojgop) [inline sdk/avalon_sdk/fabric/fabric_wrapper.py]: This function is too complex. Could be split into two - query chain code and invoke chain code which results in state change. May be in separate PR

File: sdk/avalon_sdk/fabric/unit_tests/test_fabric_worker_registry_impl.py [hunk 24]
```diff
@@ -0,0 +1,189 @@
+# Copyright 2019 Intel Corporation
```
Comment (manojgop) [inline sdk/avalon_sdk/fabric/unit_tests/test_fabric_worker_registry_impl.py]: 2020

File: sdk/avalon_sdk/fabric/unit_tests/test_fabric_worker_registry_impl.py [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+        if not path.isfile(config_file):
+            raise FileNotFoundError("File not found at path: {0}".format(
+                path.realpath(config_file)))
+        try:
+            with open(config_file) as fd:
+                self.__config = toml.load(fd)
+        except IOError as e:
+            if e.errno != errno.ENOENT:
+                raise Exception("Could not open config file: %s", e)
+        self.__eth_conn = FabricWorkerRegistryImpl(self.__config)
```
Comment (manojgop) [inline sdk/avalon_sdk/fabric/unit_tests/test_fabric_worker_registry_impl.py]: Rename __eth_conn to __fabric_conn
